From e9327d2b3c8fc3c430fb200475c175181030564d Mon Sep 17 00:00:00 2001
From: Bharathraj Nagaraju <snbraj@codeaurora.org>
Date: Thu, 26 Oct 2017 11:14:49 -0700
Subject: [PATCH 1/4] dirvers:ntb : Fix isr routine backport issue.

IDT hw dirver uses api pci_alloc_irq_vectors which
kernel 4.4 wont support. Hence using api pci_enable_msi
for msi IRQ allocation.

Change-Id: I3da6f505b2ea5709d1c49261df8244e546b46570
Signed-off-by: Bharathraj Nagaraju <snbraj@codeaurora.org>
---
 drivers/ntb/hw/idt/ntb_hw_idt.c | 33 ++++++++++++++++++++++++++++++---
 1 file changed, 30 insertions(+), 3 deletions(-)

diff --git a/drivers/ntb/hw/idt/ntb_hw_idt.c b/drivers/ntb/hw/idt/ntb_hw_idt.c
index ed806cc..4be5076 100644
--- a/drivers/ntb/hw/idt/ntb_hw_idt.c
+++ b/drivers/ntb/hw/idt/ntb_hw_idt.c
@@ -1912,6 +1912,19 @@ static int idt_init_isr(struct idt_ntb_dev *ndev)
 	u32 ntint_mask;
 	int ret;
 
+/* Enable MSI support: just enables the one MSI interrupt. Backporting NTB to 4.4 Kernel.
+   4.4 Kernel dont have support for IRQ allocator API 'pci_alloc_irq_vectors'
+*/
+#if 1
+  ret = pci_enable_msi(pdev);
+  pr_err ("%s: MSI IRQ irq = %d ", __func__, ret);
+  pr_err ("%s: MSI IRQ pdev->irq = %d ", __func__, pdev->irq);
+  if(ret)
+  {
+     dev_err(&pdev->dev, "%s:Enable MSI error\n", __func__);
+     goto err_free_vectors;
+  }
+#else
 	/* Allocate just one interrupt vector for the ISR */
 	ret = pci_alloc_irq_vectors(pdev, 1, 1, PCI_IRQ_MSI | PCI_IRQ_LEGACY);
 	if (ret != 1) {
@@ -1929,6 +1942,10 @@ static int idt_init_isr(struct idt_ntb_dev *ndev)
 	/* Set the IRQ handler */
 	ret = devm_request_threaded_irq(&pdev->dev, ret, NULL, idt_thread_isr,
 					IRQF_ONESHOT, NTB_IRQNAME, ndev);
+#endif
+	/* Set the IRQ handler */
+	ret = devm_request_threaded_irq(&pdev->dev, pdev->irq, NULL, idt_thread_isr,
+					IRQF_ONESHOT, NTB_IRQNAME, ndev);
 	if (ret != 0) {
 		dev_err(&pdev->dev, "Failed to set MSI IRQ handler, %d", ret);
 		goto err_free_vectors;
@@ -1944,8 +1961,12 @@ static int idt_init_isr(struct idt_ntb_dev *ndev)
 	return 0;
 
 err_free_vectors:
+#if 1
+  pci_disable_msi(pdev);
+  devm_free_irq(&pdev->dev, pdev->irq, ndev);
+#else
 	pci_free_irq_vectors(pdev);
-
+#endif
 	return ret;
 }
 
@@ -1965,12 +1986,18 @@ static void idt_deinit_isr(struct idt_ntb_dev *ndev)
 	ntint_mask = idt_nt_read(ndev, IDT_NT_NTINTMSK) | IDT_NTINTMSK_ALL;
 	idt_nt_write(ndev, IDT_NT_NTINTMSK, ntint_mask);
 
+#if 1
+	devm_free_irq(&pdev->dev, pdev->irq, ndev);
+#else
 	/* Manually free IRQ otherwise PCI free irq vectors will fail */
 	devm_free_irq(&pdev->dev, pci_irq_vector(pdev, 0), ndev);
-
+#endif
+#if 1
+  pci_disable_msi(pdev);
+#else
 	/* Free allocated IRQ vectors */
 	pci_free_irq_vectors(pdev);
-
+#endif
 	dev_dbg(&pdev->dev, "NTB interrupts deinitialized");
 }
 
-- 
1.9.1

